### PR TITLE
Invert error-prone configuration to be allow-list vs deny-list

### LINKED ISSUE #11969: invert error-prone configuration to be allow-list vs deny-list
### Description

Currently the error-prone is a long list of disabled checks. The problem is that on upgrade, more checks get enabled and slip into our configuration. Because of this, our error-prone has become very slow. For example android-specific checks are enabled.

The error-prone performance seems to be linear to the number of checks. I think we should disable all checks, then only enable the ones we want (e.g. ones that are not duplicated with JDK or ECJ checks). With this approach, we may be able to enable error-prone checking by default.

behavior of `main`:
```
 225.22 sec.  compileTestJava
 223.56 sec.  compileJava
```

behavior of error-prone with all checks disabled:
```
  76.24 sec.  compileJava
  64.06 sec.  compileTestJava
```

behavior of error-prone with just 12 checks enabled:
```
  84.49 sec.  compileJava
  72.10 sec.  compileTestJava
```

For this issue, i'd like to simply "invert" the list to be a list of checks that are currently enabled. So it won't really change functionality, but instead we can more visibly see what we have turned on and then fix it in a later issue.

### PR DESCRIPTION
**Pull request by rmuir**
This does not change the semantics or performance of our setup.

Instead, it explicitly enables checks that we want vs disabling checks that we don't want.

Also reordered checks to match the error-prone website list of checks for easier maintenance.

It is now clear that many useless checks are enabled, we can disable some of them and try to get the performance reasonable.

Closes #11969

### CODE REVIEW

File: gradle/validation/error-prone.gradle [hunk 1]
```diff
@@ -68,119 +68,420 @@ allprojects { prj ->
...
+            '-Xep:ClassNewInstance:WARN',
+            '-Xep:CloseableProvides:WARN',
+            // '-Xep:CollectionUndefinedEquality:OFF',
+            '-Xep:CollectorShouldNotUseState:WARN',
+            '-Xep:ComparableAndComparator:WARN',
+            '-Xep:CompareToZero:WARN',
+            // '-Xep:ComplexBooleanConstant:OFF',
+            '-Xep:DateChecker:WARN',
+            '-Xep:DateFormatConstant:WARN',
+            '-Xep:DefaultCharset:WARN',
```
Comment (madrob) [inline gradle/validation/error-prone.gradle:294]: does forbidden APIs do this one?

File: gradle/validation/error-prone.gradle [hunk 2]
```diff
@@ -68,119 +68,420 @@ allprojects { prj ->
...
-            '-Xep:UnescapedEntity:OFF',
-            '-Xep:UnicodeEscape:OFF',
-            '-Xep:UnnecessaryLambda:OFF',
-            '-Xep:UnnecessaryParentheses:OFF',
-            '-Xep:UnsynchronizedOverridesSynchronized:OFF',
-            '-Xep:UnusedMethod:OFF',
-            '-Xep:UnusedVariable:OFF',
-            '-Xep:UseCorrectAssertInTests:OFF',
-            '-Xep:WaitNotInLoop:OFF',
+            '-Xep:SwigMemoryLeak:WARN',
```
Comment (madrob) [inline gradle/validation/error-prone.gradle:445]: We don't use SWIG, could turn this off?

File: gradle/validation/error-prone.gradle [hunk 3]
```diff
@@ -68,119 +68,420 @@ allprojects { prj ->
...
+            // '-Xep:ThreadJoinLoop:OFF',
+            // '-Xep:ThreadLocalUsage:OFF',
+            // '-Xep:ThreadPriorityCheck:OFF',
+            '-Xep:ThreeLetterTimeZoneID:WARN',
+            '-Xep:TimeUnitConversionChecker:WARN',
+            // '-Xep:ToStringReturnsNull:OFF',
+            '-Xep:TruthAssertExpected:WARN',
+            '-Xep:TruthConstantAsserts:WARN',
+            '-Xep:TruthGetOrDefault:WARN',
+            '-Xep:TruthIncompatibleType:WARN',
```
Comment (madrob) [inline gradle/validation/error-prone.gradle:456]: We don't use the Truth library, could turn these off

File: gradle/validation/error-prone.gradle [hunk 4]
```diff
@@ -68,119 +68,420 @@ allprojects { prj ->
...
-            '-Xep:StaticAssignmentOfThrowable:OFF',
-            '-Xep:StaticGuardedByInstance:OFF',
-            '-Xep:StreamResourceLeak:OFF',
+            // List of enabled/disabled checks
+            // Please keep this synced with https://errorprone.info/bugpatterns when upgrading!
+
+            // On by Default : ERROR
+
+            '-Xep:AlwaysThrows:ERROR',
+            '-Xep:AndroidInjectionBeforeSuper:ERROR',
```
Comment (madrob) [inline gradle/validation/error-prone.gradle:80]: We don't use android libraries, could disable this

File: gradle/validation/error-prone.gradle [hunk 5]
```diff
@@ -68,119 +68,420 @@ allprojects { prj ->
...
+            '-Xep:FromTemporalAccessor:ERROR',
+            '-Xep:FunctionalInterfaceMethodChanged:ERROR',
+            '-Xep:FuturesGetCheckedIllegalExceptionType:ERROR',
+            '-Xep:FuzzyEqualsShouldNotBeUsedInEqualsMethod:ERROR',
+            '-Xep:GetClassOnAnnotation:ERROR',
+            '-Xep:GetClassOnClass:ERROR',
+            '-Xep:GuardedBy:ERROR',
+            '-Xep:GuiceAssistedInjectScoping:ERROR',
+            '-Xep:GuiceAssistedParameters:ERROR',
+            '-Xep:GuiceInjectOnFinalField:ERROR',
```
Comment (madrob) [inline gradle/validation/error-prone.gradle:141]: We don't use Guice

**Review by madrob — COMMENTED** [review]
There are some EP checks that are specific to libraries or frameworks that we are not using AFAIK. Could disable them or leave them in in case we start using those libraries in the future? I pointed out a few that I saw.

**Review by dweiss — APPROVED** [review]

**Comment by rmuir** [comment]
see https://github.com/apache/lucene/pull/11971

**Pull request merged**